src/inmanta/deploy/work.py

```diff
@@ -0,0 +1,424 @@
+        backing_heapq.sort()
+        return [item.task for item in backing_heapq if self._is_active(item)]
+
+    ########################################
+    # asyncio.PriorityQueue-like interface #
+    ########################################
+
+    def queue_put_nowait(self, prioritized_task: PrioritizedTask[Task]) -> None:
+        """
+        Add a new task to the associated agent's queue.
+        """
+        task: Task = prioritized_task.task
+        priority: int = prioritized_task.priority
+        already_queued: Optional[TaskQueueItem] = self._queue_item_for_task(task)
+        if already_queued is not None and already_queued.task.priority <= priority:
+            # task is already queued with equal or higher priority
+            return
+        # reschedule with new priority, no need to explicitly remove, this is achieved by setting self._tasks_by_resource
+        item: TaskQueueItem = TaskQueueItem(task=prioritized_task, insert_order=self._entry_count)
+        self._entry_count += 1
+        if task.resource not in self._tasks_by_resource:
+            self._tasks_by_resource[task.resource] = {}
+        self._tasks_by_resource[task.resource][task] = item
+        # TODO: parse agent
+        self._agent_queues["TODO"].put_nowait(item)
+
+    async def queue_get(self, agent: str) -> Task:
+        """
+        Consume a task from an agent's queue. If the queue is empty, blocks until a task becomes available.
+        """
+        queue: asyncio.PriorityQueue[TaskQueueItem] = self._agent_queues[agent]
+        while True:
+            item: TaskQueueItem = await queue.get()
+            if not self._is_active(item):
+                # task was marked as removed, ignore it and consume the next item from the queue
+                queue.task_done()
+                continue
+            # remove from the queue since it's been picked up
+            # TODO: may need to keep track of running tasks as well, but seperately because a user might want to requeue it while running
+            self.discard(item.task.task)
+            return item.task.task
+
+    def task_done(self, agent: str) -> None:
+        """
+        Indicate that a formerly enqueued task for a given agent is complete.
+
+        Used by queue consumers. For each get() used to fetch a task, a subsequent call to task_done() tells the corresponding
+        agent queue that the processing on the task is complete.
+        """
+        self._agent_queues[agent].task_done()
+
+    #########################
+    # Mapping implementation#
+    #########################
+
+    def __getitem__(self, key: Task) -> PrioritizedTask[Task]:
+        return self._tasks_by_resource.get(key.resource, {})[key].task
+
+    def __iter__(self) -> Iterator[Task]:
+        return itertools.chain.from_iterable(self._tasks_by_resource.values())
+
+    def __len__(self) -> int:
+        return sum(len(items) for items in self._tasks_by_resource.values())
```

Comment:
Looks expensive, but may not be used often?

Reply:
Not used at all atm, it's just one of the methods you need to implement the `Mapping` abc.